active/0021-transfer-files-over-mqtt.md

```diff
@@ -0,0 +1,206 @@
+
+The following terms are used as described in [MQTT Version 5.0 Specification](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc39010030):
+* Application Message
+* Server
+* Client
+* Topic Name
+* Topic Filter
+* MQTT Control Packet
+
+*At least once*: a message can be delivered many times, but cannot be lost
+
+## Requirements
+
+* The protocol MUST use only PUBLISH type of MQTT Control Packet
+* The protocol MUST support transfer of file segments
+* Server MUST be able to verify integrity of each file segment and of the whole file
+* Client MAY know total file size when initiating the transfer
+* Client MAY abort file transfer
+* Server MAY ask the client to pause file transfer
+* Server MAY ask the client to abort file transfer
+* The protocol MUST guarantee "At least once" delivery
+* Server MUST NOT support subscription on topics dedicated for file transfer
+
+## AWS IoT MQTT-based file delivery (reference design)
+
+As an example of existing implementation we can look at AWS IoT Core [which provides functionality](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery.html) to [deliver files to IoT devices](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery-in-devices.html):
+
+![](0021-assets/aws-mqtt-file-delivery.png)
+
+## Design
+
+### Overview
+
+* Files are split in segments, segments can be of arbitrary length
+* Client generates UUID for each file being transferred and use it as file Id in Topic Name
+* Client calculates sha256 checksum of the segment it's about to send and sends it as part of Topic Name
+* Client uses $file Topic Filter to transfer files
+* Clients cannot subscribe to $file topics
+* Segment length can be calculated on the server side by subtracting the length of the [Variable Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901025) from the [Remaining Length](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901105) field that is in the [Fixed Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901021)
+* Data is transferred in PUBLISH packets in the following order:
+  1. $file/{fileId}/init
+  2. $file/{fileId}/{offset}/{sha256sum}
+  3. $file/{fileId}/{offset}/{sha256sum}
+  4. ...
+  3. $file/{fileId}/[fin/{sha256sum} | abort]
+* Client can send up to N (configured) PUBLISH packets before blocking for PUBACK
+
+#### `$file/{fileId}/init` message
+
+Initialize the file transfer. Server is expected to store metadata from the payload in the session along with `{fileId}` as a reference for the rest of file metadata.
+
+  * Qos=1
+  * DUP=0 for the initial segment transfer, 1 when retransmitting
+  * Payload Format Indicator=0x01
+  * `{fileId}` is corresponding file UUID
+  * Payload is a JSON document with the following properties:
+    * filepath
+      * a character string composed with characters from this set [a-zA-Z0-9_-./]
```

Comment:
What is the reason for the character set to be limited?

Reply:
TODO: remove this requirement